### DISCUSSION TITLE
依赖注入和面向接口编程的矛盾性？

### DISCUSSION THREAD

**Discussion by Vince0911** [BYJRK/DotNet-Discussions#358] [C# 技术交流]
新手的理解，一般当提到接口特性，那是在指面向对象编程的多态特性，也就是说定义、设计了一个接口，下面要有N个对象继承它、实现它。
对于依赖注入，一般的模式是builder.Services.AddSingleton<IMessageWriter, MessageWriter>();，注入一个接口和实现它的对象。一般就一个对象去实现它（from gpt的回答），那这不是和接口的初衷正好背离？
此外如果只是一个对象去实现这个接口，那和直接注入builder.Services.AddSingleton<MessageWriter>()有何区别？我理解这时候会提到接口定义了标准，当我们修改MessageWriter内部逻辑的时候不会影响外部程序，但如果我不修改MessageWriter中的方法签名，也是一样的效果，不会对外部有影响。

**Comment by BYJRK** [comment]
其实并不矛盾，因为接口并不一定是为了多态，更多时候是为了封装，也就是只暴露你需要的成员，隐藏你不需要关注的部分。

**Comment by BYJRK** [comment]
另外，如果只有一个对象去实现这个接口，那么注入接口对象而非这个实现类型依旧是很有必要的，因为你可能需要在单元测试中对接口对象进行Mock。如果某个类的构造传入的直接就是实现类型而不是接口，那么你就不那么容易去mock了。

**Comment by hjslovehr** [comment]
简单说一下吧，面向接口编程是为了解除依赖，依赖注入是对面向接口编程的一种应用。为什么这么说呢？举个例子：A 依赖 B，那么 A 跟 B 可以说是强关联。那么怎么解除这种关系呢，可以引入接口（这里应该叫抽象），然后 A 依赖这个接口，B 实现这个接口，那么 A 和 B 就在源代码层面上互相隔离了，A 没有直接依赖 B。解除依赖可以在源代码层面上很好的实现模块划分，职责分离。但是最终代码运行还是要依赖具体的实现，那么依赖注入就是解决具体用哪个实现类的。
你可以先了解 SOLID 原则，可以看看《架构整洁之道》，需要慢慢体会。
